clippy_lints/src/string_patterns.rs

```diff
@@ -0,0 +1,239 @@
+                        && let Some(c) = get_char_value(right)
+                    {
+                        set_chars.push(c);
+                        ControlFlow::Continue(Descend::No)
+                    } else if path_to_local_id(right, binding)
+                        && cx.typeck_results().expr_ty_adjusted(left).kind() == &ty::Char
+                        && let Some(c) = get_char_value(left)
+                    {
+                        set_chars.push(c);
+                        ControlFlow::Continue(Descend::No)
+                    } else {
+                        ControlFlow::Break(())
+                    }
+                },
+                ExprKind::Binary(op, _, _) if op.node == BinOpKind::Or => ControlFlow::Continue(Descend::Yes),
+                ExprKind::Match(match_value, [arm, _], _) => {
+                    if matching_root_macro_call(cx, sub_expr.span, sym::matches_macro).is_none() {
+                        return ControlFlow::Break(());
+                    }
+                    if arm.guard.is_some() {
+                        return ControlFlow::Break(());
+                    }
+                    if !path_to_local_id(match_value, binding) {
+                        return ControlFlow::Break(());
+                    }
+                    if arm.pat.walk_short(|pat| match pat.kind {
+                        PatKind::Lit(expr) if let ExprKind::Lit(lit) = expr.kind => {
+                            if let LitKind::Char(c) = lit.node {
+                                set_chars.push(format!("'{}'", c.escape_default()));
+                            }
+                            true
+                        },
+                        PatKind::Or(_) => true,
+                        _ => false,
+                    }) {
+                        ControlFlow::Continue(Descend::No)
+                    } else {
+                        ControlFlow::Break(())
+                    }
+                },
+                _ => ControlFlow::Break(()),
+            }
+        })
+        .is_some()
+        {
+            return;
+        }
+        match set_chars.len().cmp(&1) {
+            Ordering::Equal => span_lint_and_sugg(
+                cx,
+                MANUAL_PATTERN_CHAR_COMPARISON,
+                method_arg.span,
+                "this manual char comparison can be written more succinctly",
+                "consider using a `char`",
+                set_chars[0].clone(),
+                applicability,
+            ),
+            Ordering::Greater => span_lint_and_sugg(
+                cx,
+                MANUAL_PATTERN_CHAR_COMPARISON,
+                method_arg.span,
+                "this manual char comparison can be written more succinctly",
+                "consider using an array of `char`",
+                format!("[{}]", set_chars.join(", ")),
+                applicability,
+            ),
+            Ordering::Less => {},
+        }
```

Comment:
Could use a single `span_lint_and_then` here, would also let us skip some allocation happening when the lint is allowed (the callback is not called if that is the case)